### ISSUE TITLE
clause forwarding in parallel for construct

### ISSUE THREAD

**Issue by pbrunet** [clang-omp/clang#21]
Hi,

I notice that for code like : 

``` c++
#pragma parallel for num_threads(16) firstprivate(n) lastprivate(n)
for(...)
{
}
```

clause are not correctly handle.
First, parallel and for are splited. It may not be an issue but clauses are forwarded to both when they exists. In the case above, it means:

``` c++
#pragma omp parallel firstprivate(n) num_threads(16)
#pragma omp for firstprivate(n) lastprivate(n)
for(...)
{

}
```

You should notice that `n` is firstprivate twice so that lastprivate applied on the `n` in the parallel construct and not the global one.

I don't know how it is handle for the CodeGen but it is tricky when we use clang as a library.

Thanks,
Pierrick

**Comment by alexey-bataev** [comment]
Hi Pierrik,
This is known trouble with the current implementation of OpenMP in 
clang. AST representation is very ugly(!!!) and sometimes is not quite 
correct (I mean, the source code produced by the -ast-print is not 
quite correct). CodeGen is able to handle such incorrectness in AST 
(actually, most of these bad things are implemented exactly for 
CodeGen).
Unfortunately, we have nothing to do with it for now. It requires 
redesign of the current implementation, but we don't have resources for 
this work. The correct implementation will be committed in official 
clang trunk.

Best regards,

# Alexey Bataev

Software Engineer
Intel Compiler Team
Intel Corp.

26 Февраль 2014 г. 11:25:15, pbrunet писал:

**Comment by pbrunet** [comment]
Hi Alexey,
Ok, the current AST give us the opportunity to do some thinks so it may be enought until it is released in the AST. Are you interested in others comment about AST issue or we just have to wait for the release?

Best,
Pierrick

**Comment by alexey-bataev** [comment]
Pierrick,
Yes, please send all info you have. It will help us to improve 
implementation in trunk.

Best regards,

# Alexey Bataev

Software Engineer
Intel Compiler Team
Intel Corp.

28 Февраль 2014 г. 12:15:45, pbrunet писал:

**Comment by pbrunet** [comment]
Alexey,
For now, I am working on commit : 3754f97a4eb98f3e6feff6526511ec45d878a5f6 so let me know if it is fix in a later commit. (Switch to 3.4 is a work in progress)

For now, we issued with variable in schedule clause for the chunk size. In the AST, .omp.schedule.var. is saved instead.

Also, Source Location are not easy to handle as SourceRange is from : after #pragma to the end of the line instead of the end of the captured statement.

Example:

<pre>
#pragma omp parallel
foo()
</pre>

Source location is for "omp parallel" instead of the whole code in this case.

Please ask if you want more informations.
Best,
Pierrick

**Comment by alexey-bataev** [comment]
Hi Pierrick,
Again, it is known problem. I had to do it to simplify the 
representation of combined constructs like "parallel for". I can try to 
fix the problem with SourceRegion, but not with num_threads(), if () 
and num_teams() clauses.
All these problems will be fixed in trunk.
Switch to 3.4 is complete, you can try it.

Best regards,

# Alexey Bataev

Software Engineer
Intel Compiler Team
Intel Corp.

3 Март 2014 г. 14:14:48, pbrunet писал:

**Comment by pbrunet** [comment]
Hi Alexey,

I tried the version rebased on 3.4 and it works better with atomic pragma. It is great.
When issue will be fix in trunk, it is fine, we don't need immediate modification.
I think It is a known issue but variables in "if" clauses are not correctly handle. It looks like it is already done for codegen in the AST.
Also, while I run OpenUH validation test suite, I notice that 11 tests fails while it should have only 3. Is it a known regression? (issue are with : task final, task untied, orphaned task if,  single copyprivate orphaned for schedule static and for schedule guided)

regards,
Pierrick

**Comment by alexey-bataev** [comment]
Pierrick,
I reworked AST representation for combined constructs in 23b32f7b1afed878ca04d1b8f15e3c80fbc279f2 . You can try it.

Best regards,
Alexey Bataev
 =============
Software Engineer
Intel Compiler Team
Intel Corp.

**Comment by pbrunet** [comment]
Alexey,
We will try it as soon as possible. It may need some re-factoring on our AST.rewriter so it is not for this week. I will be back when the modification is done.

Thanks for your work,
Pierrick
